v0/util/index.js

```diff
@@ -1424,10 +1426,93 @@ const getErrorStatusCode = (error, defaultStatusCode = 400) => {
 class CustomError extends Error {
   constructor(message, statusCode, metadata) {
     super(message);
+    // *Note*: This schema is being used by other endpoints like /poll, /fileUpload etc,.
+    // Apart from destination transformation
     this.response = { status: statusCode, metadata };
   }
 }
 
+/**
+ * This is the base error class which will be used as a base going forward
+ */
+ class RudderBaseError extends Error {
+  constructor(message, statusCode, statTags, destResponse, authErrorCategory) {
+    super(message);
+    // This schema will be used by transformation related endpoints
+    const errBuilder = new ErrorBuilder()
+      .setMessage(message)
+      .setStatus(statusCode)
+      .setStatTags(statTags)
+    // This would be needed for API scope
+    if (destResponse) {
+      errBuilder.setDestinationResponse(destResponse);
+    }
+    // We should basically set this for refreshing and disabling capability of OAuth destinations
+    if (authErrorCategory) {
+      errBuilder.setAuthErrorCategory(authErrorCategory);
+    }
```

Comment:
As these do not apply to all error types, they should reside in specific classes.

If we are going to use class-based approach, I suggest the following:

```
class RudderBaseError extends Error {
  constructor (message, statusCode, statTags) {
    this.message = message;
    this.status = statusCode;
    this.err.statTags = {
      destType: statTags.destType || statTags.destination,
      stage: statTags.stage,
      scope: statTags.scope,
      meta: statTags.meta
    };
  }
}

class TransformationError extends RudderBaseError {
  constructor(message, statusCode=400, statTags, destination) {
    const defScope = TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.SCOPE;
    const defMeta = TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.TRANSFORMATION.META.CLIENT_ERROR;

    let finalStatTags = statTags;
    if (!finalStatTags || Array.isArray(finalStatTags)) {
      finalStatTags = {
        scope: defScope,
        meta: defMeta
      }
    }

    const tagNames = Object.keys(finalStatTags);
    if (!tagNames.includes('scope')) {
      finalStatTags = {
        ...finalStatTags,
        scope: defScope
      }
    }

    if (!tagNames.includes('meta')) {
      finalStatTags = {
        ...finalStatTags,
        meta: defMeta
      }
    }

    finalStatTags = { ...finalStatTags, destType: destination };

    super(message, statusCode, finalStatTags);
  }
}

class ApiError extends RudderBaseError {
  constructor(message, statusCode=400, statTags, destination, destResponse, authErrCategory) {
    const defScope = TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.SCOPE;
    const defMeta = TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.META.UNHANDLED;

    let finalStatTags = statTags;
    if (!finalStatTags || Array.isArray(finalStatTags)) {
      finalStatTags = {
        scope: defScope,
        meta: defMeta
      }
    }

    const tagNames = Object.keys(finalStatTags);
    if (!tagNames.includes('scope')) {
      finalStatTags = {
        ...finalStatTags,
        scope: defScope
      }
    }

    if (!tagNames.includes('meta')) {
      finalStatTags = {
        ...finalStatTags,
        meta: defMeta
      }
    }

    finalStatTags = { ...finalStatTags, destType: destination };
    
    super(message, statusCode, finalStatTags);

    this.destinationResponse = destResponse;
    this.authErrCategory = authErrCategory;
  }
}
```

I'm sure the logic can be improved further, but you get the gist.

Reply:
addressed